File: clients/typescript/src/table.ts
import { ClientConfig, ListResponse, FindManyParams } from './types'

export class TableClient {
  constructor(private config: ClientConfig, private table: string) {}

  private url(path: string): string {
    return `${this.config.baseURL}/${this.config.app}/${path}`
  }

  private headers(): Record<string, string> {
    return {
      Authorization: `Bearer ${this.config.jwt}`,
      'Content-Type': 'application/json',
    }
  }

  private async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const res = await fetch(this.url(path), {
      method,
      headers: this.headers(),
      body: body ? JSON.stringify(body) : undefined,
    })
    if (!res.ok) {
      const err = await res.json().catch(() => ({ error: res.statusText }))
      throw new Error(err.error || `HTTP ${res.status}`)
    }
    if (res.status === 204) return undefined as T
    return res.json()
  }

  async findMany(params?: FindManyParams): Promise<ListResponse> {
    const qs = new URLSearchParams()
    if (params?.limit) qs.set('limit', String(params.limit))
    if (params?.offset) qs.set('offset', String(params.offset))
    if (params?.order) qs.set('order', params.order)
    if (params?.filters) {
      for (const [col, val] of Object.entries(params.filters)) {
        qs.set(col, val)
      }
    }
    const q = qs.toString()
    return this.request('GET', `${this.table}/${q ? `?${q}` : ''}`)
  }

  async findById(id: string): Promise<Record<string, unknown>> {
    return this.request('GET', `${this.table}/${id}`)
  }

  async create(data: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request('POST', this.table, data)
  }

  async update(id: string, data: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request('PATCH', `${this.table}/${id}`, data)
  }

  async replace(id: string, data: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.request('PUT', `${this.table}/${id}`, data)
  }

  async delete(id: string): Promise<void> {
    return this.request('DELETE', `${this.table}/${id}`)
  }
}

File: clients/go/table.go
package orbit

import (
	"fmt"
	"net/url"
)

type TableClient struct {
	client *Client
	table  string
}

func (t *TableClient) path(id string) string {
	if id == "" {
		return fmt.Sprintf("%s/", t.table)
	}
	return fmt.Sprintf("%s/%s/", t.table, id)
}

type FindManyParams struct {
	Limit   int
	Offset  int
	Order   string
	Filters map[string]string
}

func (t *TableClient) FindMany(params *FindManyParams) (*ListResponse, error) {
	q := url.Values{}
	if params != nil {
		if params.Limit > 0 {
			q.Set("limit", fmt.Sprintf("%d", params.Limit))
		}
		if params.Offset > 0 {
			q.Set("offset", fmt.Sprintf("%d", params.Offset))
		}
		if params.Order != "" {
			q.Set("order", params.Order)
		}
		for k, v := range params.Filters {
			q.Set(k, v)
		}
	}
	path := t.table
	if len(q) > 0 {
		path += "/?" + q.Encode()
	}
	var resp ListResponse
	err := t.client.request("GET", path, nil, &resp)
	return &resp, err
}

func (t *TableClient) FindByID(id string) (map[string]any, error) {
	var row map[string]any
	err := t.client.request("GET", t.path(id), nil, &row)
	return row, err
}

func (t *TableClient) Create(data map[string]any) (map[string]any, error) {
	var row map[string]any
	err := t.client.request("POST", t.table+"/", data, &row)
	return row, err
}

func (t *TableClient) Update(id string, data map[string]any) (map[string]any, error) {
	var row map[string]any
	err := t.client.request("PATCH", t.path(id), data, &row)
	return row, err
}

func (t *TableClient) Replace(id string, data map[string]any) (map[string]any, error) {
	var row map[string]any
	err := t.client.request("PUT", t.path(id), data, &row)
	return row, err
}

func (t *TableClient) Delete(id string) error {
	return t.client.request("DELETE", t.path(id), nil, nil)
}
